torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3360,11 +3360,40 @@ def _make_data_dependent_error(self, expr, unhinted_expr):
             # problem
         )
 
-    def _set_replacement(self, a: "sympy.Symbol", expr: "sympy.Expr") -> None:
+    def _set_replacement(self, a: "sympy.Symbol", expr: "sympy.Expr", msg: str) -> None:
         """
         Adds or updates a replacement for a symbol.
         Use this instead of `self.replacements[a] = expr`.
         """
+
+        # With unbacked symbols, it is not necessarily always profitable to do
+        # a replacement.  We may end up with a worse value range if we do the
+        # substitution, especially if the unbacked symbol to be replaced was
+        # marked as size like.
+        #
+        # For example let's suppose that 'a' in [2, inf].  If we replace 'a'
+        # with '10 - y', where y in [2, inf], we get a much worse bound.  This
```

Comment:
Sorry, I forgot to say what the final bound here is, which is `[-inf, 8]`. It's true that the bounds are not strictly comparable, but the new bound doesn't know that 10 - y is positive which is bad.

Reply:
Ah, so you want to more aggressively refine bounds on the source variables. In this case that would work, because these are linear functions. What about when this is not possible? Maybe it doesn't matter!

Reply:
I am not sure what it means for a variable to have bounds of the form `[2, inf]`, all I'm saying is that this is a simple transformation we can do and it shouldn't be stopping us from implementing any analysis.

Reply:
I am thinking about this more and I am worried about doing this refinement as is.

Consider the following situation: y has compile-time bounds `[2, 3]` and run-time bounds `[0, 3]` (it's size-like.) What are the compile time bounds for `x = 3 - y`? If we know x is size-like then we have `x in [2, inf]`, solving backwards, this would imply that `y in [-inf, 1]`. But when we intersect the two bounds, we find y is unsatisfiable!

This is a consequence of unsound 0/1 specialization. I talk about it more at https://github.com/pytorch/pytorch/issues/117361 Right now, we are mostly side stepping the problem because we are not doing analysis that is too "strong". But if we add this backward propagation, I fear we will run into. I'd prefer not to at this point in time.

Reply:
I'm probably missing something, but you can't have a variable with compile-time bounds of `[2,3]` and run-time bounds of `[0,3]`. The compile-time bound must be an over-approximation of the observed values at run time.

Reply:
It's unsound :)

Reply:
Is there any path forward for avoiding unsoundness ? having to deal with it is a continual headache and I think we're just going to keep on papering on fixes.

Maybe just a way of encapsulating it. It's a very local context that makes it hard to reason throughout the codebase

Reply:
@eellison, I mean, what do you think of my proposal in https://github.com/pytorch/pytorch/issues/117361 ?

Reply:
> Is there any path forward for avoiding unsoundness ? having to deal with it is a continual headache and I think we're just going to keep on papering on fixes.
> 
> Maybe just a way of encapsulating it. It's a very local context that makes it hard to reason throughout the codebase

What is "it" here?

Reply:
I will continue discussion on that thread. 

it: unsoundness. i.e. maintaining a range of >=2 for something that in runtime has a range of >=2.